### PR TITLE
An attempt to fix onPress not being called when pressing near element boundaries

### LINKED ISSUE #1061: Press events are not dispatched correctly on iOS Chrome/Safari
# 🐛 Bug Report 

Press events are not dispatched correctly on iOS.

I've created a `Button` component with the `useButton` hook and I noticed press events are not dispatched correctly (but click events are) on iOS Chrome/Safari (I didn't test on Android) if you click in any of the Button edges. It works on desktop though.

## 🤔 Expected Behavior

`useButton` / `usePress` should dispatch press events every time a click event would be dispatched.

## 😯 Current Behavior

Screencast showing the problem on Chrome iOS: https://d.pr/v/5haGuQ

## 💻 Code Sample

Codesandbox link: https://codesandbox.io/s/sad-meitner-ssc9k

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| @react-aria/button   | 3.2.1
| Browser          | Chrome 85.0.4183.92
| Operating System | iOS 13.6.1

### PR DESCRIPTION
**Pull request by Andarist**
Closes #1061

Problem nicely illustrated here: https://github.com/adobe/react-spectrum/issues/1061#issuecomment-689917900

⚠️ This is just an initial attempt and a signal that I'd like to fix this and I'm willing to contribute the bugfix. Currently, there are issues with this code.

Initially, I've thought that something like this should work because I've thought that the whole problem lies in the `isOverTarget` logic but the problem seems to be a little bit more complicated. The main complexity comes IMHO from the fact that you are using both local~ and global events to cover this functionality because of reasons ™.

I understand this mainly tries to workaround some browser inconsistencies and such but given that I don't have the full context behind this, it's way harder for me to figure out the whole fix on my own. If you have a lit of problems that you have encountered when working on this and the reasoning behind current implementation written something down I would highly appreciate a link to it so I could study this more.

My current reasoning is that this version that I have **won't** work mainly because of 2 buttons close to each other as this `isOverTarget` call is in the global `onPointerUp` handler so basically a "sibling" element can be triggered using this logic. So IMHO to handle this correctly using the current approach we'd have to loop through all registered elements, deconflict stuff, and basically replicate browser behaviors that are not really that well documented. This is not something that I'd like to pursue as this doesn't scale and raise the complexity tremendously.

To the best of my understanding, the "local" `onPointerUp` is the best place to determine if a particular press event is targeting the target element. This is resolved by the browser and at least we shouldn't receive any other `event.currentTarget` there than expected. From what I got from some comments left in the code there are some cases for which this is called incorrectly in Safari on iOS from time to time but I'm wondering - why not just ignore those cases that are incorrect and handle the rest in there? I feel like I might be missing some context here.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

**Comment by Andarist** [comment]
Actually - we have tested this and it works OK for us. I've analyzed the code thoroughly and I don't see how this could fail. Only the element that receives a starting event (like pointerdown) is tracked through global listeners, so while this approach is a little bit quirky (I understand that u probably had good reasons to use it though) it shouldn't actually cause any unwanted `onPress` calls as far as I can tell and it was the thing that I was worried about.

**Comment by devongovett** [comment]
Thanks @Andarist and sorry for the delay. We'll need to test this very carefully as usePress is used in many components. Do you also mind adding a unit test as well to cover these cases?

File: packages/@react-aria/interactions/src/usePress.ts [hunk 1]
```diff
@@ -602,15 +602,54 @@ function createEvent(target: HTMLElement, e: EventBase): EventBase {
...
+  clientY: number,
+  width?: number,
+  height?: number,
+  radiusX?: number,
+  radiusY?: number
+}
+
+function getPointClientRect(point: EventPoint): PointRect {
+  let offsetX = (point.width / 2) || point.radiusX || 0;
+  let offsetY = (point.height / 2) || point.radiusY || 0;
```
Comment (devongovett) [inline packages/@react-aria/interactions/src/usePress.ts:628]: Do we need to check if `point.width`/`point.height` are `null` here? Feels weird to rely on NaN coercion behavior...

Comment (Andarist) [reply packages/@react-aria/interactions/src/usePress.ts:628]: Hah, I was actually quite surprised that TS has allowed this. I completely agree that relying on NaN coercion feels weird but at the same i've chosen this as other versions included more nesting and such and looked more complicated in the end than it really is.

Maybe I should just add a comment above those lines that this relies on NaN coercion?

**Comment by Andarist** [comment]
> We'll need to test this very carefully as usePress is used in many components. 

Understandable. I would actually still be interested in learning more why this local/global event listeners dance is implemented here if you don't mind. I wonder what weird browser edge cases you are handling there 😅 maybe it would be good to write this down in a comment or smth? I bet most contributors/readers would be interested in that.

> Do you also mind adding a unit test as well to cover these cases?

Sure thing. I could also add some actual browser-based test but I dont recall you having here any Cypress/Puppeteer/* setup.

Gonna work on this after the weekend.

**Comment by Andarist** [comment]
@devongovett I've added some unit tests for this and improved the implementation slightly

**Comment by Andarist** [comment]
@devongovett friendly 🏓 about this PR :)

**Review by devongovett — APPROVED** [review]

**Review by dannify — APPROVED** [review]

**Comment by devongovett** [comment]
Thanks @Andarist, and apologies again for the long delay on this PR. We are having a testing session tomorrow and will test across a wide variety of devices, but I tested locally and it seems to work well for me.

**Comment by nicosammito** [comment]
This specific issue still exists for me
